drv/stm32h7-update-server/src/main.rs

```diff
@@ -0,0 +1,270 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+//
+// Functions for writing to flash for updates
+//
+// This driver is intended to carry as little state as possible. Most of the
+// heavy work and decision making should be handled in other tasks.
+#![no_std]
+#![no_main]
+
+use drv_update_api::UpdateError;
+use ringbuf::*;
+use stm32h7::stm32h753 as device;
+use userlib::*;
+
+const FLASH_KEY1: u32 = 0x4567_0123;
+const FLASH_KEY2: u32 = 0xCDEF_89AB;
+
+const FLASH_OPT_KEY1: u32 = 0x0819_2A3B;
+const FLASH_OPT_KEY2: u32 = 0x4C5D_6E7F;
+
+const BANK_ADDR: u32 = 0x08100000;
+const BANK_END: u32 = 0x08200000;
+
+// Writes are indexed by flash words, BANK_ADDR is word 0,
+// BANK_ADDR + FLASH_WORD_BYTES is word 1 etc.
+const BANK_WORD_LIMIT: usize =
+    (BANK_END - BANK_ADDR) as usize / FLASH_WORD_BYTES;
+
+// RM0433 Rev 7 section 4.3.9
+// Flash word is defined as 256 bytes
+const FLASH_WORD_BITS: usize = 256;
+
+// Total length of a word in bytes (i.e. our array size)
+const FLASH_WORD_BYTES: usize = FLASH_WORD_BITS / 8;
+
+// number of 32-bit words we will have to write
+const FLASH_CPU_WORDS: usize = FLASH_WORD_BYTES / 4;
+
+// Block is an abstract concept here. It represents the size of data the
+// driver will process at a time.
+const BLOCK_SIZE: usize = FLASH_WORD_BYTES * 32;
```

Comment:
This is a little confusing – maybe define
```rust
const FLASH_WORDS_PER_BLOCK: usize = 32;
```
first, then define `BLOCK_SIZE_BYTES` in terms of it (instead of just `BLOCK_SIZE`, which is ambiguously unitless)?

Reply:
Hmmm, I don't think that makes it clearer. `FLASH_WORDS_PER_BLOCK` is closer to an internal constant than something that we actually want to change (see where it gets used). `BLOCK_SIZE` is what we would actually want to adjust but I'll see about changing that to `BLOCK_SIZE_BYTES` to indicate the unit.